builder/build_validate.go

```diff
@@ -273,100 +293,178 @@ func (b *Builder) mcaPkgInfo(manifests []*swupd.Manifest, version, downloadRetri
+					resolveList = append(resolveList, p)
+				}
+				bundlePkgInfo.allPkgs[p.name] = pkg
+			}
 
-		for pkg := range m.BundleInfo.AllPackages {
-			pkgList = append(pkgList, pkg)
-		}
+			// TODO: resolveRpmURIs is a bottleneck when all files are local. Running this function with a long
+			// list of packages is significantly faster than running many times with small lists.
 
-		out, err := downloadRpms(packagerCmd, pkgList, buildVersionDir, downloadRetries)
-		if err != nil {
-			return nil, err
-		}
+			// Resolve the URI of each new rpm
+			if err := resolveRpmURIs(resolveList, repo, repoURIs[repo], pkgInfoCache, packagerCmd); err != nil {
+				return nil, err
+			}
 
-		// Collect metadata to resolve installed RPM file names
-		pInfo[m.Name].allPkgs, err = pkgInfoFromNoopInstall(out.String())
-		if err != nil {
-			return nil, errors.Wrapf(err, m.Name)
-		}
-		for _, p := range pInfo[m.Name].allPkgs {
-			if pkgFileCache[p.name] == nil {
-				pkgFileCache[p.name], err = b.resolvePkgFiles(p, version)
-				if err != nil {
+			// Query file metadata for each new package
+			for _, p := range resolveList {
+				pkg, ok := pkgInfoCache[p.name]
+				if !ok {
+					return nil, errors.Errorf("Failed to load pkg %s from map", p.name)
+				}
+				select {
+				case pkgCh <- pkg:
+				case err := <-errCh:
 					return nil, err
 				}
 			}
-			p.files = pkgFileCache[p.name]
+		}
+	}
+	close(pkgCh)
+	wg.Wait()
 
-			// Track all resolved package files in bundle to be used later
-			// for file subtraction.
+	if len(errCh) > 0 {
+		return nil, <-errCh
+	}
+
+	// Create map of files in each bundle
+	for _, bundlePkgInfo := range pInfo {
+		for _, p := range bundlePkgInfo.allPkgs {
 			for _, f := range p.files {
-				pInfo[m.Name].allFiles[f.name] = true
+				bundlePkgInfo.allFiles[f.name] = true
 			}
 		}
 	}
 
-	return pInfo, err
+	return pInfo, nil
 }
 
 // resolvePkgFiles queries a package for a list of file metadata.
-func (b *Builder) resolvePkgFiles(pkg *pkgInfo, version int) ([]*fileInfo, error) {
+func (b *Builder) resolvePkgFiles(pkg *pkgInfo, downloadRetries int) ([]*fileInfo, error) {
+	var err error
+	var out *bytes.Buffer
+
 	queryCmd := "[%{filenames}\a%{filesizes}\a%{filedigests}\a%{filemodes:perms}\a%{filelinktos}\a%{fileusername}\a%{filegroupname}\n]"
 	rpmCmd := []string{"rpm", "-qp", "--qf=" + queryCmd}
 
-	validationDir := filepath.Join(b.Config.Builder.ServerStateDir, "validation")
-	buildVersionDir := filepath.Join(validationDir, fmt.Sprint(version))
-	pkgDir := filepath.Join(buildVersionDir, "packages")
-
-	pkgFileName := pkg.name + "-" + pkg.version + "." + pkg.arch + ".rpm"
-	pkgPath := filepath.Join(pkgDir, pkgFileName)
-
 	// Query RPM for file metadata lists
-	args := merge(rpmCmd, pkgPath)
-	out, err := helpers.RunCommandOutputEnv(args[0], args[1:], []string{"LC_ALL=en_US.UTF-8"})
-	if err != nil {
-		return nil, err
+	args := merge(rpmCmd, pkg.uri)
+	for attempts := 0; attempts <= downloadRetries; attempts++ {
```

Comment:
won't this run 1 additional time?

Reply:
I thought of it as the number of times to try again. So, 3 retries would be 1 attempt + 3 retries.

Reply:
aaah.. makes sense.